cryostatio/cryostat-operator
# feat(auth): add config option to deploy oauth2_proxy instead of openshift-oauth-proxy

# Welcome to Cryostat! 👋
## Before contributing, make sure you have:
_______________________________________________

Related to #710
Fixes #206

## Description of the change:
This enables the Operator to deploy `oauth2_proxy` in place of `openshift-oauth-proxy`.

`openshift-oauth-proxy` is deployed by default when the Operator detects that it is installed on OpenShift. There is a new CRD option to disable this integration and force the use of `oauth2_proxy` instead. If installed on a non-OpenShift Kubernetes then `oauth2_proxy` will always be deployed.

If `oauth2_proxy` is deployed it will also respect the same htpasswd Basic auth configuration that I added in #802. If no htpasswd Basic auth is configured then `oauth2_proxy` will allow any requests to pass without authz.

## Motivation for the change:
1. enable auth proxy functionality for installs on non-OpenShift clusters
2. enable end users to deploy Cryostat with htpasswd Basic, with or without OpenShift SSO login - they may not want to use cluster RBAC to control access to the Cryostat API and application, but rather to define separate htpasswd user accounts. Or they may wish to provide access to users who have cluster accounts as well as some users who don't
3. enable end users to deploy Cryostat without any authentication required at all. The configuration is locked down so integrating with other OAuth providers is not currently possible (not via the Operator anyway), but for quick one-off testing or small/private testing clusters, configuring auth may be an unwanted additional step

## How to manually test:
1. Build and install Operator as usual
2. Create a CR:

```yaml
apiVersion: operator.cryostat.io/v1beta2
kind: Cryostat
metadata:
  name: cryostat-sample
spec:
  enableCertManager: false
  authorizationOptions:
    disableOpenShiftSso: false
  trustedCertSecrets: []
  eventTemplates: []
  storageOptions:
    pvc:
      labels: {}
      annotations: {}
      spec: {}
  reportOptions:
    replicas: 0
```

3. ensure that the application comes up and is secured by OpenShift SSO as before
4. create a basic auth secret like in #802 (`oc create secret generic basicauth --from-file=htpasswd.conf`) patch the CR to add:

```yaml
authorizationOptions:
  basicAuth:
    secretName: basicauth
    filename: htpasswd.conf
```

5. ensure that the application is redeployed and is now secured by OpenShift SSO, with the option of Basic auth as well
6. edit the CR to set `disableOpenShiftSso: true`
7. ensure that the application is redeployed and is now secured by OAuth2 Proxy using the same Basic auth
8. edit the CR to remove the `basicAuth` object
9. ensure that the application is redeployed and is now accessible without auth
10. bonus: install operator on a non-OpenShift cluster. The `disableOpenShiftSso` flag should now be ineffective and OAuth2 Proxy should be installed regardless. The `basicAuth` config option should still function as outlined in the previous steps.
